pkg/services/hcloud/server/server.go

```diff
@@ -190,6 +217,198 @@ func (s *Service) handleBootStateUnset(ctx context.Context) (reconcile.Result, e
+		s.scope.SetError(msg, capierrors.CreateMachineError)
+		return reconcile.Result{}, nil
+	}
+
+	if hm.Status.ActionIDEnableRescueSystem != actionDone {
+		action, err := s.scope.HCloudClient.GetAction(ctx, hm.Status.ActionIDEnableRescueSystem)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("GetAction failed: %w", err)
+		}
+
+		if action.Finished.IsZero() {
+			log.FromContext(ctx).Info("Waiting until Action RescueEnabled is finished",
+				"action", action)
+			return reconcile.Result{RequeueAfter: requeueIntervalWaitForPreRescueOSThenEnableRescueSystem}, nil
+		}
+
+		err = action.Error()
+		if err != nil {
+			err = fmt.Errorf("action %+v failed: %w", action, err)
+			log.FromContext(ctx).Error(err, "")
+			s.scope.SetError(err.Error(), capierrors.CreateMachineError)
+			return reconcile.Result{}, nil
+		}
+
+		log.FromContext(ctx).Info("Action RescueEnabled is finished",
+			"actionDuration", action.Finished.Sub(action.Started),
+			"finishedSince", time.Since(action.Finished))
+
+		hm.Status.ActionIDEnableRescueSystem = actionDone
+	}
+
+	action, err := s.scope.HCloudClient.Reboot(ctx, server)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("reboot failed: %w", err)
+	}
+
+	hm.Status.ActionIDRebootToRescue = action.ID
+
+	hm.SetBootState(infrav1.HCloudBootStateWaitForRescueRunningThenInstallImage)
+	return reconcile.Result{RequeueAfter: requeueAfterRebootToRescue}, nil
+}
+
+func (s *Service) handleBootStateWaitForRescueRunningThenInstallImage(ctx context.Context, _ *hcloud.Server) (reconcile.Result, error) {
+	hm := s.scope.HCloudMachine
+
+	if hm.Status.ActionIDRebootToRescue == 0 {
+		msg := "handleBootStateWaitForRescueRunningThenInstallImage ActionIdRebootToRescue not set? Can not continue. Provisioning Failed"
+		log.FromContext(ctx).Error(nil, msg)
+		s.scope.SetError(msg, capierrors.CreateMachineError)
+		return reconcile.Result{}, nil
+	}
+	if hm.Status.ActionIDRebootToRescue != actionDone {
+		action, err := s.scope.HCloudClient.GetAction(ctx, hm.Status.ActionIDRebootToRescue)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("GetAction failed: %w", err)
+		}
+
+		if action.Finished.IsZero() {
+			log.FromContext(ctx).Info("Waiting until Action RebootToRescue is finished", "action", action)
+			return reconcile.Result{RequeueAfter: requeueIntervalWaitForRescueRunning}, nil
+		}
+
+		err = action.Error()
+		if err != nil {
+			err = fmt.Errorf("action %+v failed: %w", action, err)
+			log.FromContext(ctx).Error(err, "")
+			s.scope.SetError(err.Error(), capierrors.CreateMachineError)
+			return reconcile.Result{}, nil
+		}
+
+		log.FromContext(ctx).Info("Action RebootToRescue is finished",
+			"actionDuration", action.Finished.Sub(action.Started),
+			"finishedSince", time.Since(action.Finished))
+
+		hm.Status.ActionIDRebootToRescue = actionDone
+	}
+
+	robotSecretName := s.scope.HetznerCluster.Spec.SSHKeys.RobotRescueSecretRef.Name
+	if robotSecretName == "" {
```

Comment:
in this case we would have to set a condition, this is a permanent thing

Reply:
fixed in 2e0a2ce3